### DISCUSSION TITLE
The transcoder does not fully utilize the CPU.

### DISCUSSION THREAD

**Discussion by Bistiara** [OvenMediaLabs/OvenMediaEngine#1647] [Help] [answered]
Good afternoon,
OME is running in a container - airensoft/ovenmediaengine:0.16.5
The virtual machine has the following specifications: 64GB RAM/32 vCPU (AMD EPYC 9654), network 1Gbps.

The problem is that no matter what settings I use for the transcoder, after about 10-12 streams, I start getting queue errors, which leads to RAM overflow, and eventually, everything crashes due to OOM (Out of Memory). I have tried setting various values for Profile, Preset, and ThreadCount, but nothing helps. What's most interesting is that the CPU is only utilized up to 20%, and I can't manage to make it use more. Please advise what I am doing wrong?
<img width="1256" alt="image" src="https://github.com/AirenSoft/OvenMediaEngine/assets/12187418/1f3b5c13-b24a-4055-9ef0-1cf5233d7524">
[logs+conf.zip](https://github.com/user-attachments/files/15838288/logs%2Bconf.zip)

**Comment by bchah** [comment]
I think software decode is bound to a single thread per instance in OME. I hit this issue all the time with Thumbnail publishers and the only way to prevent OME from consuming all system memory is to use the experimental [DeleteLazyStream](https://github.com/AirenSoft/OvenMediaEngine/discussions/1221) feature. This is still not ideal because it would stop the transcode even when there is plenty of CPU available.

Is there any difference if you run this against the master branch? I believe they changed the decoder to use to 2 threads in commit 9e9be09.

**Reply by Bistiara** [reply]
Returned to the problem, unfortunately, there is no difference. Currently on version 0.16.8, the CPU is still only being used at 20%. I tried it on a virtual machine with 64GB RAM/32 vCPU (AMD EPYC 9654) and also tried from the master branch, but the behavior remains the same.

**Comment by getroot** [comment]
Could you please check which threads need to be distributed using top -H -p command?
Then capture it and upload it so I can figure out what settings you need.

https://airensoft.gitbook.io/ovenmediaengine/performance-tuning#monitoring-the-usage-of-threads

**Reply by Bistiara** [reply]
Good afternoon, I ran 2 tests on a virtual machine with 64 GB RAM and 64 vCPUs (AMD EPYC 9654), handling 12 streams. After some time, everything crashes due to memory overflow. The top, htop outputs, metrics, logs, and configuration files are in the archive. Thank you in advance for your response.
[oven_test.zip](https://github.com/user-attachments/files/16608747/oven_test.zip)

**Reply by getroot** [reply]
After checking the data you sent, it seems like the bottleneck is in the video filter. I will add this to the backlog so that we can work on it together with the task of setting the number of decoder threads in the future.

**Reply by Bistiara** [reply]
So, do I understand correctly that on the SW decoder, regardless of the processor's power, we won't be able to run more than 10-12 streams? Are there any possible ways to run more?

**Reply by getroot** [reply]
It's not the decoder, but the video filter is the bottleneck in your system as mentioned above. I haven't seen a case where the video filter is the bottleneck yet, so I need to do some research. (The video filter runs on 1 thread) I'm going to check the system performance using your Server.xml.

And the encoding performance is not the number of inputs, but the output. In your case, it's 30-36 because you output 3 per video input.

First, to make the video filter work as little as possible, try setting the output profile to the same fps and resolution as the input. In your case, it's 29.97 and 1440x1080. The height is currently omitted in the output, but it would be better to specify the height as well.

**Reply by Bistiara** [reply]
Good afternoon, thank you very much for taking the time to this issue. Unfortunately, changing the frame rate and resolution does not seem to have resolved the problem ( Width:1440,Height:1080,Framerate:29.97). I have attached the configuration and logs in the archive.
[ome_test3.zip](https://github.com/user-attachments/files/16656849/ome_test3.zip)

**Reply by Keukhan** [reply]
@Bistiara 

I checked the problem you reported. The reason is that multi-threading is operating differently than expected in OpenH264 encoders.

One approach is to effectively utilize the process CPU by changing the scheduling policy of the process to RR, FIFO. However, there is a problem that it cannot be used without root user. so, I have a plan to replace the H264 encoder with x264. It is expected that the problem will be solved if it is replaced. 

I can't replace it because I don't have enough time right now, but I'll make a plan. 
Wait for the good news.

Cheers.

**Reply by Bistiara** [reply]
Thank you for your explanation @Keukhan @getroot . We are open to implementing a temporary solution. Could you provide more details on where exactly we can change the process scheduling policy to RR or FIFO? In which part of the system or code can this be done, and what specific steps need to be taken?

**Reply by Bistiara** [reply]
Good day, on this issue, perhaps you need some additional information from us? so that you could tell me what temporary solution can be used?

**Reply by Keukhan** [reply, marked as answer]
@Bistiara 


First, I fixed the issue with thread usage not being accurate. I also added the x264 library. 
In short, OpenH264 doesn't show the exact number of threads it creates. It looks like all the encoders share the same threads.  but, the H264 decoder, scaling filter, and x264 encoder seem to handle CPU thread usage well. 

To use x264, use `<Codec>h264_x264</Codec>`

**Patches**
- https://github.com/AirenSoft/OvenMediaEngine/commit/f37ed418a244c81e629d658484e579d5fc30e830
- https://github.com/AirenSoft/OvenMediaEngine/commit/76e918f8c2e082362c1334d8548ab9a18d1bbca4

**CPU load balance test result**
- https://airensoft.atlassian.net/wiki/external/Y2E1NWNkYjc3ZjcwNGMxNWI5M2ZjNmEzZDE0YzY0NDA

The patched code has not been released yet. The next version will be released soon. thanks.
